compiler/p/codegen/OMRTreeEvaluator.cpp

```diff
@@ -5902,114 +5902,350 @@ OMR::Power::TreeEvaluator::generateHelperBranchAndLinkInstruction(
+      {
+      TR::Register *valCopyReg = cg->allocateRegister();
       generateTrg1Src1Instruction(cg, TR::InstOpCode::mr, valueNode, valCopyReg, valueReg);
+
       valueReg = valCopyReg;
-      stopUsingCopyReg3 = true;
+      stopUsingCopyRegVal = true;
       }
 
    TR::LabelSymbol * residualLabel =  generateLabelSymbol(cg);
    TR::LabelSymbol * loopStartLabel =  generateLabelSymbol(cg);
    TR::LabelSymbol * doneLabel =  generateLabelSymbol(cg);
-   TR::LabelSymbol * label8aligned =  generateLabelSymbol(cg);
-   TR::LabelSymbol * label4aligned =  generateLabelSymbol(cg);
-   TR::LabelSymbol * label2aligned =  generateLabelSymbol(cg);
-   TR::LabelSymbol * label1aligned =  generateLabelSymbol(cg);
+
+   //these labels are not needed for the vector approach to storing to residual bytes (i.e.: P10+)
+   TR::LabelSymbol *label8aligned, *label4aligned, *label2aligned, *label1aligned;
+
+   if (!cg->comp()->target().cpu.isAtLeast(OMR_PROCESSOR_PPC_P10))
+      {
+      label8aligned =  generateLabelSymbol(cg);
+      label4aligned =  generateLabelSymbol(cg);
+      label2aligned =  generateLabelSymbol(cg);
+      label1aligned =  generateLabelSymbol(cg);
+      }
 
    TR::RegisterDependencyConditions *conditions;
-   int32_t numDeps = 5;
+   int32_t numDeps = 6;
+
+   //need extra register for offset only if it isn't already included in the destination address AND it isn't a constant
+   if (arrayCheckNeeded && !useOffsetAsImmVal)
+      numDeps++;
+
    conditions = new (cg->trHeapMemory()) TR::RegisterDependencyConditions(numDeps, numDeps, cg->trMemory());
    TR::Register *cndReg = cg->allocateRegister(TR_CCR);
    TR::addDependency(conditions, cndReg, TR::RealRegister::cr0, TR_CCR, cg);
-   TR::addDependency(conditions, dstAddrReg, TR::RealRegister::NoReg, TR_GPR, cg);
+
+   if (arrayCheckNeeded)
+      {
+      TR::addDependency(conditions, dstBaseAddrReg, TR::RealRegister::NoReg, TR_GPR, cg);
+
+      if (!useOffsetAsImmVal)
+         TR::addDependency(conditions, dstOffsetReg, TR::RealRegister::NoReg, TR_GPR, cg);
+      }
+   else
+      {
+      TR::addDependency(conditions, dstAddrReg, TR::RealRegister::NoReg, TR_GPR, cg);
+      }
+
    TR::addDependency(conditions, lengthReg, TR::RealRegister::NoReg, TR_GPR, cg);
    TR::addDependency(conditions, valueReg, TR::RealRegister::NoReg, TR_GPR, cg);
-   TR::Register * tempReg = cg->allocateRegister();
-   TR::addDependency(conditions, tempReg, TR::RealRegister::NoReg, TR_GPR, cg);
+   TR::Register * temp1Reg = cg->allocateRegister();
+   TR::Register * temp2Reg = cg->allocateRegister();
+   TR::addDependency(conditions, temp1Reg, TR::RealRegister::NoReg, TR_GPR, cg);
+   TR::addDependency(conditions, temp2Reg, TR::RealRegister::NoReg, TR_GPR, cg);
+
+   //need to exlude GPR0 from the following registers:
+   //1.) dstBaseAddrReg/destAddrReg, since it holds the address of the object being written to
+   //2.) temp1Reg, since it will later hold the J9Class flags for the object at dstAddr
```

Comment:
this is error-prone in the future, if anybody changed the order of adding dependency.  look around for other examples of dependency build-up ... usually we do gpr0 exclusion at the point of adding that particular dependency and using the dependency index retrieved from conditions itself.